Only check access permissions in ``/api/{history_dataset_collection_id}/contents/{dataset_collection_id}``

ahmedhamidawan opened the pull request:
Only check ownership in `__get_history_collection_instance` if:
`collection_instance.history.published` = False

Fixes https://github.com/galaxyproject/galaxy/issues/17252
Closes https://github.com/galaxyproject/galaxy/issues/15917

### -
_Is this currently the right way to go about this or do we also need to add a `published` param to the api as well to add context?..._

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
fix published history collections are empty on expand

Only check ownership in `__get_history_collection_instance` if:
`collection_instance.history.published` = False

mvdbeek commented on lib/galaxy/managers/collections.py:
```diff
@@ -802,7 +802,7 @@ def __get_history_collection_instance(
         # TODO: that sure looks like a bug, we can't check ownership using the history of the object we're checking ownership for ...
         history = getattr(trans, "history", collection_instance.history)
-        if check_ownership:
+        if check_ownership and collection_instance.history.published is False:
```
That is what `check_accessible=True` is supposed to do, `check_ownership` is the check that is applied if we're doing stuff for which you need to be owner. I think you need to find the API routes that are problematic, and then check if we're modifying the collection, which means `check_ownership=True`, or if we're just viewing then `check_accessible=True` is fine

ahmedhamidawan replied:
Ah i see, that makes a lot of sense, I should instead look at whether any modifications are being made or if we're just viewing, and check ownership/accessibility accordingly..

ahmedhamidawan replied:
Updated the changes, does this make more sense?: I removed the `check_ownership=True` flag since this api (`/api/dataset_collections/{hdca_id}/contents/{parent_id}`) doesn't modify a collection. This also works as it allows you to access collections from public histories and doesn't if history is private.

If this is right, will also work on fixing L803 in `lib/galaxy/managers/collections.py` here https://github.com/galaxyproject/galaxy/pull/17444/commits/1e8e88556b2df0870d57ea84b4c6256a58e434fd#diff-b58c9c9fec7966bb1f9faf0ec34cb50a8474a0180290ee7fd9accae0923a9597R803

ahmedhamidawan committed:
remove `check_ownership` from `dataset_collections.contents`

... since this api only needs to `check_accessible` as it is not modifying a collection

mvdbeek reviewed (approved):
Yes, looks good, thank you! Any chance you could add an API test ?
First part of that test could just look like https://github.com/mvdbeek/galaxy/blob/b7652623b424506363183393795e851476574856/lib/galaxy_test/api/test_dataset_collections.py#L413-L420, and then publish the history with `self.dataset_populatior.make_public(history_id)` and check that a different user can list the contents ?

ahmedhamidawan committed:
add api test for hdcas in public histories being accessible
